pymanda/choices.py

```diff
@@ -213,11 +222,231 @@ def restrict_data(self, restriction):
+            
+            df_shares = df_shares.rename(columns = {weight_var: 'share'})
+            output_dict.update({key: df_shares})
+
+        return output_dict
+    
+    def shares_checks(self, df, share_col, data="Data"):
+        """ Checks for share columns"""
+        
+        if share_col not in df.columns:
+            raise KeyError("Column '{}' not in ChoiceData")
+        if (df[share_col] < 0).any():
+            raise ValueError ("Values of '{col}' in {d} contain negative values".format(col=share_col, d=data))
+        if df[share_col].sum() != 1:
+            raise ValueError ("Values of '{col}' in {d} do not sum to 1".format(col=share_col, d=data))
+    
+    def calculate_hhi(self, shares_dict, share_col="share", group_col=None):
+        """
+        Calculates HHIs from precalculated shares at the corporation level
+        
+        Parameters
+        ----------
+        shares_dict: dictionary of share tables
+            
+        share_col : Column name in dataframe, Optional
+            column that holds float of shares. Default is 'share'
+        
+        group_col: Column name in dataframe, Optional
+            column of names to calculate HHI on. Default is self.corp_var
+
+        Returns
+        -------
+        Dictionary
+            keys are the same as shares_dict
+            values are hhis
+        
+        Examples
+        --------
+        >>> corps = ['x' for x in range(50)]
+        >>> corps += ['y' for x in range(25)]
+        >>> corps += ['z' for x in range(25)]
+        
+        >>> choices = ['a' for x in range(30)]
+        >>> choices += ['b' for x in range(20)]
+        >>> choices += ['c' for x in range(20)]
+        >>> choices += ['d' for x in range(5)]
+        >>> choices += ['e' for x in range(25)]
+        >>> df = pd.DataFrame({"corporation": corps,
+                          "choice" : choices})
+        >>> cd = ChoiceData(df, "choice", corp_var="corporation")
+        
+        >>> shares = cd.calculate_shares()
+        >>> cd.calculate_hhi(shares)
+        {'Base Shares': 3750.0}
+
+        """
+        if type(shares_dict) != dict:
+            raise TypeError ("Expected type dict. Got {}".format(type(shares_dict)))
+        
+        if group_col is None:
+            group_col = self.corp_var
+        elif group_col not in self.data.columns:
+            raise KeyError ('''"{}" is not a column in ChoiceData'''.format(group_col))
+        output_dict = {}
+        for key in shares_dict.keys():
+            df = shares_dict[key]
+            
+            if type(df) != pd.core.frame.DataFrame:
+                raise TypeError ('''Expected type pandas.core.frame.DataFrame Got {}'''.format(type(df)))
+            
+            self.shares_checks(df, share_col, data=key)
+            
+            df = df.groupby(group_col).sum()
+            df[share_col] = df[share_col] * 100
+            hhi = (df[share_col] * df[share_col]).sum()
```

Comment:
i'd prefer to do this part as:
```suggestion
            hhi = (df[share_col] * df[share_col]).sum()*10000
```